src/charm.py

```diff
@@ -260,6 +269,189 @@ def _on_remove(self, event: RemoveEvent):
                     pass
             self._delete_vault_data()
 
+    def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent):
+        """Handler triggered when a new vault-kv client is attached."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can configure a vault-kv client, skipping")
+            return
+
+        if not self._is_peer_relation_created():
+            logger.debug("Peer relation not created, deferring event")
+            event.defer()
+            return
+
+        try:
+            root_token, _ = self._get_initialization_secret_from_peer_relation()
+        except PeerSecretError:
+            logger.debug("Vault initialization secret not set in peer relation, deferring event")
+            event.defer()
+            return
+
+        try:
+            (
+                _,
+                _,
+                ca_certificate,
+            ) = self._get_certificates_secret_in_peer_relation()
+        except PeerSecretError:
+            logger.debug("Vault certificate secret not set in peer relation, deferring event")
+            event.defer()
+            return
+
+        relation = self.model.get_relation(event.relation_name, event.relation_id)
+
+        if relation is None or relation.app is None:
+            logger.warning(
+                "Relation or remote application is missing,"
+                "this should not happen, skipping event"
+            )
+            return
+
+        vault = Vault(url=self._api_address)
+        vault.set_token(token=root_token)
+
+        if not vault.is_api_available():
+            logger.debug("Vault is not available, deferring event")
+            event.defer()
+            return
+
+        vault.enable_approle_auth()
+
+        mount = "charm-" + relation.app.name + "-" + event.mount_suffix
+        vault.configure_kv_mount(mount)
+        self.vault_kv.set_mount(relation, mount)
+        vault_url = self._get_relation_api_address(relation)
+        if vault_url is not None:
+            self.vault_kv.set_vault_url(relation, vault_url)
+        self.vault_kv.set_ca_certificate(relation, ca_certificate)
+
+        nonces = []
+        for unit in relation.units:
+            egress_subnet = relation.data[unit].get("egress_subnet")
+            nonce = relation.data[unit].get("nonce")
+            if egress_subnet is None or nonce is None:
+                logger.debug(
+                    "Skipping configuring access for unit %r, egress_subnet or nonce are missing",
+                    unit.name,
+                )
+                continue
+            nonces.append(nonce)
+            self._ensure_unit_credentials(vault, relation, unit.name, mount, nonce, egress_subnet)
+
+        # Remove any stale nonce
+        credential_nonces = self.vault_kv.get_credentials(relation).keys()
+        stale_nonces = list(set(credential_nonces) - set(nonces))
```

Comment:
If we change the type hint in the library, this would do:
```suggestion
        stale_nonces = set(credential_nonces) - set(nonces)
```